### DISCUSSION TITLE
Sending JSON object to datadog with Vercel log drains integration

### DISCUSSION THREAD

**Discussion by yoan-elba** [vercel/community#139] [Show & Tell]
I use Datadog log drains integration, when I log a JSON object inside my handler, Vercel take my JSON and add some informations like 
billed duration, memory size ...

But it sent all of the informations as a text in message field, so i can't use Datadog to make filter on my own log data or categorise my log using a custom error level, see the screenshot below. 

<img width="720" alt="example" src="https://user-images.githubusercontent.com/95749791/146806091-ba21391b-74b5-4e2b-9c5f-844b41f72411.png">

**Comment by Pieparker** [comment]
Hey there! Is this still an issue for you? It may be worth asking the Datadog team for their thoughts. I'll also ask internally, in case any of our team have dealt with this before.

**Comment by yoan-elba** [comment]
For my side i will workaround by removing vercel data dog integration and
make the integration by myself.
But I really don't understand which project can use your datadog
integration because actually it's just impossible to find any information
about my log application. Your integration send only plain text to datadog,
why not send a json format ?

**Reply by Pieparker** [reply]
For information regarding Datadog's integration and whether parsing the data as JSON is possible, I'd definitely consider reaching out to them–especially if their integration is your first choice.

So you're aware, [Logflare](https://logflare.app/guides/vercel-setup) have documented how they automatically parse text messages and is another viable solution for log draining.

**Reply by hydrosquall** [reply]
Parsing the data as JSON is possible. JSON is usually auto-detected, but it doesn't work in the case of the screenshot because some text appears in front of the JSON body. If `Version` and `Start requestId` were keys in the object instead of on their own line, automatic parsing would work.

Unless the logging format changes on the Vercel side, your other option is to write a Grok parser for Datadog that will either keep the fields from the first line (or drop the first line), and parse the rest as JSON.

https://docs.datadoghq.com/logs/log_configuration/parsing/

**Comment by rezanouri87** [comment]
This problem makes the integration quite difficult to use. Plaintext log in datadog is useless. Is this something that will get fixed?

**Comment by davidbfrogman** [comment]
I was able to fix this by changing the grok parsing rule to this:

`report_rule .*\s+RequestId: %{notSpace:lambda.request_id}%{word:version}\s+Version: \$LATEST\s+%{data::json}\s+Duration: %{number:lambda.duration:scale(1000000)} ms\s+Billed Duration: %{number:lambda.billed_duration} ms\s+Memory Size: %{number:lambda.memorysize} MB\s+Max Memory Used: %{number:lambda.max_memory_used} MB(\s+Init Duration: %{number:lambda.init_duration} ms)?.*`

This might not work for everyone, but it worked for me. I used pino to log the messages as json, and the grok parsing rule to pull the json out of the message in the middle of the text.

The key bit in there is: `%{data::json}`

**Reply by Thinkscape** [reply]
This only works if your lambda emits a single log entry ... if there's multiple json objects, the whole thing fails.

**Reply by davit-b** [reply]
For a single object, the slightly modified Grok parser rule worked for me
```
report_rule .*\s+RequestId: %{notSpace:lambda.request_id}%{word:version}\s+Version: \$LATEST\s+%{data:analytics:json}\s+Duration: %{number:lambda.duration:scale(1000000)} ms\s+Billed Duration: %{number:lambda.billed_duration} ms\s+Memory Size: %{number:lambda.memorysize} MB\s+Max Memory Used: %{number:lambda.max_memory_used} MB(\s+Init Duration: %{number:lambda.init_duration} ms)?.*
```

Note I changed `%{data::json}` --> `%{data:analytics:json}` so that all my custom data is nested under a key called `analytics` to distinguish it.

<img width="961" alt="Screenshot 2023-07-07 at 2 48 31 PM" src="https://github.com/vercel/community/assets/12516060/c999e2c4-e51d-4dd4-8808-60dbd9465d32">

@Thinkscape If there are multiple JSON objects, can the key-value parser pattern be applied? https://docs.datadoghq.com/logs/log_configuration/parsing/?tab=matchers#key-value-or-logfmt

**Reply by Thinkscape** [reply]
Nah mate, you can't use key-value parser. It only applies key values to a single log entry.
What we need is a log entry **splitter**, or a first-party JSONL support, or any reasonable log format support... or ability to opt out of the lambda noise decoration.

**Reply by davit-b** [reply]
Thank you, I learned a lot just googling those things.

I'm going to consider creating a custom log drain
Vercel -> Lambda ( log entry splitter) -> Datadog

**Reply by Thinkscape** [reply]
This might get extremely expensive with volume (you're multiplying your bill for every `n` requests), and would be quite wasteful all things considered. I believe a fix would be trivial for Vercel/ @vercel-infra if they were willing to address this long-standing issue 🤷  Because Vercel wraps the AWS lambda, and they already ingest the logs (from cloudwatch I presume) then they could easily split it and provide a clean log drain instead of the current spaghetti ...

**Comment by tebuevd** [comment]
Any updates on this? @vercel-infra tagging you folks for visibility

**Comment by dankain** [comment]
Is there an update on this please? Logging is very important to us in resolving issues with serverless functions. To us each log event is a structured data JSON object, each event being separate. This then allows us in datadog to search for unique events or errors. Vercel is concatenating all log lines from one lambda request and sending them as a string to datadog. This makes them almost unusable to us in datadog. @vercel-infra do you have a plan to fix this please? Without this visibility we would be unable to migrate some of our existing serverless functionality into Vercel. I see the same concatenated logging in the Vercel dashboard, so assume this is how they are being consumed from cloudwatch
